src/gpu/nodeKernels.cu

```diff
@@ -229,25 +229,32 @@ __global__ void kProcessBeamSamplesFirstLast(size_t beamCount, int samplesPerBea
 	LIMIT(beamCount);
 
 	const auto beamIdx = tid;
-	int firstIdx = 0;
-	int lastIdx = 0;
+	int firstIdx = -1;
+	int lastIdx = -1;
 	for (int sampleIdx = 0; sampleIdx < samplesPerBeam; ++sampleIdx) {
 		if (beamSamples.isHit[beamIdx * samplesPerBeam + sampleIdx] == 0) {
 			continue;
 		}
-		if (beamSamples.distance[beamIdx * samplesPerBeam + sampleIdx] <
-		    beamSamples.distance[beamIdx * samplesPerBeam + firstIdx]) {
+		auto currentFirstDistance = firstIdx >= 0 ? beamSamples.distance[beamIdx * samplesPerBeam + firstIdx] : FLT_MAX;
+		auto currentLastDistance = lastIdx >= 0 ? beamSamples.distance[beamIdx * samplesPerBeam + lastIdx] : -FLT_MAX;
+		if (beamSamples.distance[beamIdx * samplesPerBeam + sampleIdx] < currentFirstDistance) {
 			firstIdx = sampleIdx;
 		}
-		if (beamSamples.distance[beamIdx * samplesPerBeam + sampleIdx] >
-		    beamSamples.distance[beamIdx * samplesPerBeam + lastIdx]) {
+		if (beamSamples.distance[beamIdx * samplesPerBeam + sampleIdx] > currentLastDistance) {
 			lastIdx = sampleIdx;
 		}
 	}
-	first.xyz[beamIdx] = beamSamples.xyz[beamIdx * samplesPerBeam + firstIdx];
-	first.distance[beamIdx] = beamSamples.distance[beamIdx * samplesPerBeam + firstIdx];
-	last.xyz[beamIdx] = beamSamples.xyz[beamIdx * samplesPerBeam + lastIdx];
-	last.distance[beamIdx] = beamSamples.distance[beamIdx * samplesPerBeam + lastIdx];
+	first.isHit[beamIdx] = firstIdx >= 0;
+	if (first.isHit[beamIdx]) {
+		first.xyz[beamIdx] = beamSamples.xyz[beamIdx * samplesPerBeam + firstIdx];
+		first.distance[beamIdx] = beamSamples.distance[beamIdx * samplesPerBeam + firstIdx];
```

Comment:
Regarding our discussion from today (setting both first and last at the main ray, not on side rays) are these xyz here in LiDAR space or in world? If this is a local space, and if I can assume that main ray is at index 0 between samplesPerBeam, then it could be possible to scale xyz of main ray hit point using its own distance and the distances of first and last.

Reply:
On the other hand, adding self-transform to MultiReturnPointers (if obviously possible at all), would solve the case for world space. I guess, it is something to maybe discuss further.

Reply:
Unfortunately, `xyz` is in world CF here.

> On the other hand, adding self-transform to MultiReturnPointers (if obviously possible at all), would solve the case for world space.

I think this could work, however cleanliness of such solution is very questionable, since it looses precision by converting back-and-forth. Let's wait and see if it's required to implement.